### ISSUE TITLE
RFE: provide an option to forbid migrating processes to a different unit/cgroup

### ISSUE THREAD

**Issue by lnykryn** [systemd/systemd#6356]
- Request for enhancement (RFE)

We see quite a lot of problems caused by incorrect use of PAM or by calling su commands in initscripts.
The problem is, that if you start your deamon from initscript via su call, systemd pam module will move it to a session scope. The biggest problem then comes with shutdown, where the sessions scopes are terminated pretty early with simply sigterm/sigkill.

Even this is a bug in those services, we should provide a way for users to start those services correctly.

So my proposal is to introduce a new option, that would force that processes will remain in the service unit (cgroup) and won't be migrated to a different one.

**Comment by Werkov** [comment]
> Even this is a bug in those services, we should provide a way for users to start those services correctly.

I thought that this was already done in `man su` and `man runuser` where it suggests to use `setpriv` in such cases (a root switching to another user and PAM session not required). (I understand the service files should be rewritten in nicer manner anyway.)



> So my proposal is to introduce a new option

What option? To `su(do)`, service/slice/scope directive prohibiting the migration or something else?

**Comment by arvidjaar** [comment]
@lnykryn 
> I thought that this was already done ... use `setpriv`

The problem is legacy initscripts used by third party software that are outside of control of administrator.

> What option?

I do not remember whether I posted it, but idea I had was additional session class, e.g. `service`. `pam_systemd` already supports using `XDG_SESSION_CLASS` to force it; so we can add `service` class that will skip re-scoping of session and make unit use `Environment=XDG_SESSION_CLASS=service`. Actually intention was for initscript generator to always add this automatically.

As far as I can tell required changes are minimal.

**Comment by JDM1492** [comment]
Part of the behavior looks like a bug but might be fixed with the control suggested.  Here is what I've encountered -- I login using a display manager and the DM gets assigned to the logged-in user's CGroup:

     user-1000.slice - User Slice of UID 1000
     Loaded: loaded
    Drop-In: /lib/systemd/system/user-.slice.d
             └─10-defaults.conf
     Active: active since Sat 2022-02-26 13:23:37 EST; 6h ago
       Docs: man:user@.service(5)
      Tasks: 47 (limit: 9921)
     Memory: 100.2M
     CGroup: /user.slice/user-1000.slice
             ├─session-1719.scope
             │ ├─65584 /usr/bin/slim -nodaemon
             │ ├─65631 /bin/sh - /home/jdm/.xinitrc
             │ ├─65646 dbus-launch --sh-syntax --exit-with-x11
             │ ├─65647 /usr/bin/dbus-daemon --syslog --fork --print-pid 5  --session
             └─user@1000.service
                     ├─dbus.service
                     │ └─64038 /usr/bin/dbus-daemon --session --address=systemd: --nofork --nopidfile --systemd-activa...
                     ├─init.scope
                     │ ├─64010 /lib/systemd/systemd --user
                     │ └─64012 (sd-pam)
                     └─pulseaudio.service
                        └─65839 /usr/bin/pulseaudio --daemonize=no --log-target=journal

Slim is the display-manager.service and runs as root. It doesn't exit at logout.  It forks an xinit task as the entering user and waits for it to exit then allows another login.  Since Slim never exits, the user is never marked as completely logged out; so users services are not explicitly stopped.  Most user services die of natural causes, but they don't get respawned when the same user logs in again because this is not a new session to systemd.    Specifically, pulseaudio doesn't start; and sound does not work for the new X11 session.

If Slim exits or kills itself at logout, everything works fine because systemd restarts it.  However, a logged-out session is different from an initial session; so some functionality in Slim is lost.  

My current workaround is in slim.conf:
`     sessionstop_cmd	systemctl restart slim
`
Possibly Slim and not systemd is broken, but there must be a good number of broken programs that would start to work if systemd could be told not to migrate certain processes to a different CGroup.  

Mostly I don't see why a process not owned by a user should be in its slice.  Maybe this issue needs to be advanced to bug.

**Comment by Werkov** [comment]
> Mostly I don't see why a process not owned by a user should be in its slice. 

I agree (under interpretation "owned by" ~ "waited for"). Then why does the display manager call into PAM (pam_systemd more precisely) and puts non-terminating process into the session?
